src/main/java/io/strimzi/kafka/access/SecretDependentResource.java

```diff
@@ -72,49 +75,44 @@ public Map<String, String> desired(final KafkaAccessSpec spec, final String name
-        final Map<String, String> data  = new HashMap<>(commonSecretData);
-        final KafkaListener listener;
-        try {
-            final String kafkaUserType = Optional.ofNullable(kafkaUser.getSpec())
+            kafkaUserType = Optional.ofNullable(kafkaUser.getSpec())
                     .map(KafkaUserSpec::getAuthentication)
                     .map(KafkaUserAuthentication::getType)
                     .orElse(KafkaParser.USER_AUTH_UNDEFINED);
+            data.putAll(getKafkaUserSecretData(context, kafkaUser, kafkaUserName, kafkaUserNamespace));
+        }
+        try {
             listener = KafkaParser.getKafkaListener(kafka, spec, kafkaUserType);
-            data.putAll(kafkaUserData.getConnectionSecretData());
         } catch (CustomResourceParseException e) {
             throw new IllegalStateException("Reconcile failed due to ParserException " + e.getMessage());
         }
+        if (listener.isTls()) {
+            listener.withCaCertSecret(getKafkaCaCertData(context, kafkaClusterName, kafkaClusterNamespace));
+        }
         data.putAll(listener.getConnectionSecretData());
         return data;
     }
 
+    private Map<String, String> getKafkaUserSecretData(final Context<KafkaAccess> context, final KafkaUser kafkaUser, final String kafkaUserName, final String kafkaUserNamespace) {
+        final String userSecretName = Optional.ofNullable(kafkaUser.getStatus())
+                .map(KafkaUserStatus::getSecret)
+                .orElseThrow(illegalStateException("Secret in KafkaUser status", kafkaUserNamespace, kafkaUserName));
+        final InformerEventSource<Secret, KafkaAccess> kafkaUserSecretEventSource = (InformerEventSource<Secret, KafkaAccess>) context.eventSourceRetriever()
+                .getResourceEventSourceFor(Secret.class, KafkaAccessReconciler.KAFKA_USER_SECRET_EVENT_SOURCE);
+        final Secret kafkaUserSecret = kafkaUserSecretEventSource.get(new ResourceID(userSecretName, kafkaUserNamespace))
+                .orElseThrow(illegalStateException(String.format("Secret %s for KafkaUser", userSecretName), kafkaUserNamespace, kafkaUserName));
+        return new KafkaUserData(kafkaUser).withSecret(kafkaUserSecret).getConnectionSecretData();
+    }
+
+    private Map<String, String> getKafkaCaCertData(final Context<KafkaAccess> context, String kafkaClusterName, String kafkaClusterNamespace) {
+        final String caCertSecretName = KafkaResources.clusterCaCertificateSecretName(kafkaClusterName);
+        final InformerEventSource<Secret, KafkaAccess> strimziSecretEventSource = (InformerEventSource<Secret, KafkaAccess>) context.eventSourceRetriever()
+                .getResourceEventSourceFor(Secret.class, KafkaAccessReconciler.STRIMZI_SECRET_EVENT_SOURCE);
+        return strimziSecretEventSource.get(new ResourceID(caCertSecretName, kafkaClusterNamespace))
+                .map(Secret::getData)
+                .orElse(new HashMap<>());
```

Comment:
Unless this method is required to return a mutable map, perhaps `Map.of()` would be better here?

Reply:
Good point, I'll change it.